### PR TITLE
HHH-14230 Fix generics to avoid ClassCastException

### PR DESCRIPTION
**Pull request by quaff**
https://hibernate.atlassian.net/browse/HHH-14230

### CODE REVIEW

File: hibernate-core/src/main/java/org/hibernate/mapping/Table.java [hunk 1]
```diff
@@ -50,7 +50,7 @@
 	/**
 	 * contains all columns, including the primary key
 	 */
-	private Map columns = new LinkedHashMap();
+	private Map<String, Column> columns = new LinkedHashMap<String, Column>();
```
Comment (NathanQingyangXu) [inline hibernate-core/src/main/java/org/hibernate/mapping/Table.java:53]: We don't need to duplicate generics spec on the right due to diamond operator. Other than that, this PR seems good. A good example we need to pay more attention to generics usage, partly because of legacy reason, partly because of coding sloppiness.

**Review by NathanQingyangXu — CHANGES_REQUESTED** [review]
I will approve the PR if you delete unnecessary generics spec on right.

File: hibernate-core/src/main/java/org/hibernate/mapping/Table.java (hunk 1 above)
Comment (quaff) [reply hibernate-core/src/main/java/org/hibernate/mapping/Table.java:53]: I'd prefer diamond operator, but it breaks consistency with legacy code, It’s necessary to cleanup entire codebase in another commit.

**Comment by dreab8** [comment]
Hi @quaff thanks for the PR, can you please also provide a test for the issue?

**Comment by quaff** [comment]
> Hi @quaff thanks for the PR, can you please also provide a test for the issue?

Coming soon.

File: hibernate-core/src/main/java/org/hibernate/mapping/Table.java [hunk 2]
```diff
@@ -420,10 +420,9 @@ public void validateColumns(Dialect dialect, Mapping mapping, TableMetadata tabl
 				throw new HibernateException( "Missing column: " + col.getName() + " in " + Table.qualify( tableInfo.getCatalog(), tableInfo.getSchema(), tableInfo.getName()));
 			}
 			else {
-				final boolean typesMatch =
-						dialect.equivalentTypes( columnInfo.getTypeCode(), col.getSqlTypeCode( mapping ) )
-						|| col.getSqlType( dialect, mapping ).toLowerCase(Locale.ROOT)
-								.startsWith( columnInfo.getTypeName().toLowerCase(Locale.ROOT) );
+				final boolean typesMatch = col.getSqlType( dialect, mapping ).toLowerCase(Locale.ROOT)
+						.startsWith( columnInfo.getTypeName().toLowerCase(Locale.ROOT) )
+						|| columnInfo.getTypeCode() == col.getSqlTypeCode( mapping );
```
Comment (dreab8) [inline hibernate-core/src/main/java/org/hibernate/mapping/Table.java:425]: Why this change?

File: hibernate-core/src/main/java/org/hibernate/mapping/Table.java (hunk 1 above)
Comment (dreab8) [reply hibernate-core/src/main/java/org/hibernate/mapping/Table.java:53]: I agree with @NathanQingyangXu , better to use the diamond operator.

**Comment by dreab8** [comment]
> Coming soon.

Thanks

File: hibernate-core/src/main/java/org/hibernate/mapping/Table.java (hunk 1 above)
Comment (quaff) [reply hibernate-core/src/main/java/org/hibernate/mapping/Table.java]: done.

File: hibernate-core/src/main/java/org/hibernate/mapping/Table.java (hunk 2 above)
Comment (quaff) [reply hibernate-core/src/main/java/org/hibernate/mapping/Table.java]: Fixed changes by accidental formatting.

**Review by NathanQingyangXu — APPROVED** [review]

**Comment by NathanQingyangXu** [comment]
I've approved this PR. I just have one open question here: is it preferable to name testing class by ticket number or descriptive  natural language? I saw coexistence of both in our codebase. Personally I always opt for the latter, though sometimes I found I am facing trouble to come up with terse name restrictive to the ticket's scope.

**Comment by quaff** [comment]
> I've approved this PR. I just have one open question here: is it preferable to name testing class by ticket number or descriptive natural language? I saw coexistence of both in our codebase. Personally I always opt for the latter, though sometimes I found I am facing trouble to come up with terse name restrictive to the ticket's scope.

Naming is difficult, especially for those people don't speak english, you should understand.

**Comment by NathanQingyangXu** [comment]
> > I've approved this PR. I just have one open question here: is it preferable to name testing class by ticket number or descriptive natural language? I saw coexistence of both in our codebase. Personally I always opt for the latter, though sometimes I found I am facing trouble to come up with terse name restrictive to the ticket's scope.
> 
> Naming is difficult, especially for those people don't speak english, you should understand.

One of the benefits of natural language test class name is for searching by keyword. It is invaluable if someone wanna search to see whether some similar bugs have been fixed.

**Comment by Sanne** [comment]
Thanks! merged

**Comment by NathanQingyangXu** [comment]
@quaff , I have started an ambitious PR to improve the generics usage in our codebase in v6: https://github.com/hibernate/hibernate-orm/pull/3587. It might take a couple of weeks or even months to have it done.
